userdetails-cognito/src/main/groovy/au/org/ala/userdetails/CognitoUserService.groovy

```diff
@@ -858,4 +867,78 @@ class CognitoUserService implements IUserService<UserRecord, UserPropertyRecord,
         resultStreamer.complete()
     }
 
+    @Override
+    Map generateApikey(String usagePlanId) {
+        if(!usagePlanId){
+            return [apikeys:null, err: "No usage plan id to generate api key"]
+        }
+
+        CreateApiKeyRequest request = new CreateApiKeyRequest()
+        request.enabled = true
+        request.customerId = currentUser.userId
+        request.name = "API key for user " + currentUser.userId
+        def response = apiGatewayIdp.createApiKey(request)
+
+        if(isSuccessful(response)) {
+            //add api key to usage plan
+            CreateUsagePlanKeyRequest usagePlanKeyRequest = new CreateUsagePlanKeyRequest()
+            usagePlanKeyRequest.keyId = response.id
+            usagePlanKeyRequest.keyType = "API_KEY"
+            usagePlanKeyRequest.usagePlanId = usagePlanId
+            apiGatewayIdp.createUsagePlanKey(usagePlanKeyRequest)
+
+            return [apikeys:getApikeys(currentUser.userId), error: null]
+        }
+        else{
+            return [apikeys:null, error: "Could not generate api key"]
+        }
+    }
+
+    @Override
+    def getApikeys(String userId) {
+
+        GetApiKeysRequest getApiKeysRequest = new GetApiKeysRequest().withCustomerId(userId).withIncludeValues(true)
+        GetApiKeysResult response = apiGatewayIdp.getApiKeys(getApiKeysRequest)
+        if(isSuccessful(response)){
+            return response.items.value
+        }
+        else{
+            return null
+        }
+    }
+
+    @Override
+    def generateClient(String userId, List<String> callbackURLs, boolean forGalah){
+        CreateUserPoolClientRequest request =  new CreateUserPoolClientRequest().withUserPoolId(poolId)
+        request.clientName = "Client for user " + userId
+        request.allowedOAuthFlows = ["code"]
+        request.generateSecret = false
+        request.supportedIdentityProviders = ["COGNITO", "Facebook", "Google", "AAF", "SignInWithApple"]
+        request.preventUserExistenceErrors = "ENABLED"
+        request.explicitAuthFlows = ["ALLOW_REFRESH_TOKEN_AUTH", "ALLOW_CUSTOM_AUTH", "ALLOW_USER_SRP_AUTH", "ALLOW_USER_PASSWORD_AUTH"]
```

Comment:
This is probably ok but just in case lets make it configurable as well.

Reply:
Fixed